modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala

```diff
@@ -17,25 +17,87 @@
 package smithy4s.codegen
 package internals
 
-import io.circe.Codec
-import io.circe.generic.semiauto._
+import cats.syntax.all._
+import io.circe._
 import io.circe.syntax._
+import io.circe.generic.semiauto._
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.openapi.OpenApiConfig
+
+import scala.collection.Seq
+import scala.reflect.ClassTag
+import scala.util.Try
 
 private[internals] final case class SmithyBuild(
     version: String,
-    imports: Seq[String],
-    maven: SmithyBuildMaven
-)
+    imports: Seq[os.FilePath],
+    plugins: Seq[SmithyBuildPlugin],
+    maven: Option[SmithyBuildMaven]
+) {
+  def getPlugin[T <: SmithyBuildPlugin](implicit
+      classTag: ClassTag[T]
+  ): Option[T] =
+    plugins.collectFirst { case t: T => t }
+}
+
 private[codegen] object SmithyBuild {
-  implicit val codecs: Codec[SmithyBuild] = deriveCodec
-  def writeJson(sb: SmithyBuild): String = sb.asJson.spaces4
+  // automatically map absence of value to empty Seq for ease of use
+  implicit def optionalSeqDecoder[T](implicit
+      base: Decoder[T]
+  ): Decoder[Seq[T]] =
+    Decoder.decodeOption(Decoder.decodeSeq[T]).map(_.getOrElse(Seq.empty))
```

Comment:
I had to decide between having a semantically correct model that's easy to use with complex parsing logic, or a model closer to the actual SmithyBuild JSON file in structure that was easy to parse but hard to use. I opted for the first, which means:
* custom `Seq` decoder so the model doesn't have to use `Option[Seq[T]]` to handle missing fields (I believe circe-generic-extras actually has a configuration option for this, but it doesn't seem to be available in all the Scala versions we have here so I went for the homebrewed version)
* modelling the plugins as a sequence, with a hand-written decoder that reads an object of the form
```{
   "openapi": (openapi config structure goes here),
   "other-plugin": (other plugin structure goes here)
}
```
and looks up the appropriate decoder for each plugin. (Although right now you could model this as a case class with a field `openApi` instead, plugin names can be dynamic so that version wouldn't be future-proof).

Let me know if you think this is too much Circe magic.

Reply:
Nah I'm fine with it, thank you 👍